File: Demo/InternationalDemo/Objective-C/BUADVADemo/ExampleApp/SubPage/Native/BUDNativeExampleView.m
//
//  BUDNativeExampleView.m
//  BUADVADemo
//
//  Created by bytedance on 2020/11/11.
//  Copyright © 2020 bytedance. All rights reserved.
//

#import "BUDNativeExampleView.h"

@interface BUDNativeExampleView ()
@property (weak, nonatomic) IBOutlet UILabel *titleLabel;
@property (weak, nonatomic) IBOutlet UILabel *subTitleLabel;
@property (weak, nonatomic) IBOutlet UIImageView *imageView;
@property (weak, nonatomic) IBOutlet UIButton *customButton;
@property (nonatomic, strong) BUNativeAdRelatedView *nativeAdRelatedView;
@end

@implementation BUDNativeExampleView

- (void)setNativeAd:(BUNativeAd *)nativeAd {
    _nativeAd = nativeAd;
    
    _titleLabel.text = _nativeAd.data.AdTitle;
    _subTitleLabel.text = _nativeAd.data.AdDescription;
    [_customButton setTitle:_nativeAd.data.buttonText?:@"CustomButton" forState:UIControlStateNormal];
    
    _nativeAdRelatedView = [[BUNativeAdRelatedView alloc] init];
    [_nativeAdRelatedView refreshData:_nativeAd];
    if (_nativeAdRelatedView.videoAdView) {
        [self addVideoView];
    }
    ///add logo view
    [self addLogoView];
    ///dislike view
    [self addDislikeView];
    ///regist view
    [_nativeAd registerContainer:self withClickableViews:@[self.customButton]];
    ///load image
    __weak typeof(self) weakSelf = self;
    NSString *imageUrl = _nativeAd.data.imageAry.firstObject.imageURL;
    if (imageUrl) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
            NSURLSessionDataTask *task = [session dataTaskWithURL:[NSURL URLWithString:imageUrl] completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
                if (data) {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        weakSelf.imageView.image = [UIImage imageWithData:data];
                    });
                }
            }];
            [task resume];
        });
    }
}
- (CGFloat)getHeight:(CGFloat)width {
    CGFloat height = 140;
    CGFloat scale = (_nativeAd.data.imageAry.firstObject.height?:0)/(_nativeAd.data.imageAry.firstObject.width?:1);
    return width*scale + height;
}
///ExampleUI
- (void)addVideoView {
    UIView *videoView = _nativeAdRelatedView.videoAdView;
    videoView.translatesAutoresizingMaskIntoConstraints = NO;
    [_imageView addSubview:videoView];
    [_imageView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-0-[videoView]-0-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(videoView)]];
    [_imageView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-0-[videoView]-0-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(videoView)]];
}
- (void)addLogoView {
    UIView *logo = _nativeAdRelatedView.logoADImageView;
    logo.translatesAutoresizingMaskIntoConstraints = NO;
    [_imageView addSubview:logo];
    [_imageView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:[logo(40)]-0-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(logo)]];
    [_imageView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:[logo(15)]-0-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(logo)]];
}
- (void)addDislikeView {
    UIButton *dislikeView = _nativeAdRelatedView.dislikeButton;
    dislikeView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:dislikeView];
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:[dislikeView(20)]-20-|" options:0 metrics:nil views:NSDictionaryOfVariableBindings(dislikeView)]];
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-20-[dislikeView(20)]" options:0 metrics:nil views:NSDictionaryOfVariableBindings(dislikeView)]];
}

@end

File: Demo/InternationalDemo/Swift/BUADVADemo_Swift/ExampleApp/SubPage/BUDNativeExampleView.swift
//
//  BUDNativeView.swift
//  BUADVADemo_Swift
//
//  Created by bytedance on 2020/11/8.
//

import UIKit

@IBDesignable class BUDNativeExampleView: UIView {

    @IBOutlet weak var titleLabel: UILabel!
    @IBOutlet weak var subTitleLabel: UILabel!
    @IBOutlet weak var cusButton: UIButton!
    @IBOutlet weak var imageView: UIImageView!
    
    private var nativeAdRelatedView:BUNativeAdRelatedView?
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        self.backgroundColor = .systemBackground
    }
    
    var nativeAd: BUNativeAd? {
        didSet {
            if nativeAd==nil {
                return
            }
            ///UI
            titleLabel.text = nativeAd?.data?.adTitle;
            subTitleLabel.text = nativeAd?.data?.adDescription;
            cusButton.setTitle(nativeAd?.data?.buttonText ?? "CustomButton", for: .normal)
            
            ///RelatedView
            nativeAdRelatedView = BUNativeAdRelatedView.init();
            nativeAdRelatedView?.refreshData(nativeAd!)
            if nativeAdRelatedView?.videoAdView != nil {
                ///add video view
                addVideoView()
            }
            ///add logo view
            addLogoView()
            ///dislike view
            addDislikeView()
            
            ///regist view
            nativeAd?.registerContainer(self, withClickableViews: [self.cusButton])
            
            ///load image
            DispatchQueue.global().async {
                let session = URLSession.init(configuration: URLSessionConfiguration.default)
                let task = session.dataTask(with: NSURL.init(string: self.nativeAd?.data?.imageAry.first?.imageURL ?? "")! as URL) { (data, respose, error) in
                    if data != nil {
                        DispatchQueue.main.async {
                            self.imageView.image = UIImage.init(data: data!)
                        }
                    }
                }
                task.resume();
            }
        }
    }
    
    func getHeight(width:Float) -> Float {
        let height = 140;
        let scale = Float(nativeAd?.data?.imageAry.first?.height ?? 0)/Float(nativeAd?.data?.imageAry.first?.width ?? 1);
        return width*scale + Float(height);
    }
    
    
    
}
///Example UI
extension BUDNativeExampleView {
    func addVideoView() -> Void {
        let videoView:UIView = nativeAdRelatedView?.videoAdView ?? UIView.init()
        videoView.translatesAutoresizingMaskIntoConstraints = false
        imageView.addSubview(videoView)
        imageView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|-0-[videoView]-0-|", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["videoView":videoView]))
        imageView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-0-[videoView]-0-|", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["videoView":videoView]))
    }
    func addLogoView() -> Void {
        let logo:UIImageView = nativeAdRelatedView?.logoADImageView ?? UIImageView.init()
        logo.translatesAutoresizingMaskIntoConstraints = false
        imageView.addSubview(logo)
        imageView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:[logo(40)]-0-|", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["logo":logo]))
        imageView.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:[logo(15)]-0-|", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["logo":logo]))
    }
    func addDislikeView() -> Void {
        let dislikeView:UIButton = nativeAdRelatedView?.dislikeButton ?? UIButton.init()
        dislikeView.translatesAutoresizingMaskIntoConstraints = false
        self.addSubview(dislikeView)
        self.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:[dislikeView(20)]-20-|", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["dislikeView":dislikeView]))
        self.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|-20-[dislikeView(20)]", options: NSLayoutConstraint.FormatOptions.init(), metrics: nil, views: ["dislikeView":dislikeView]))
    }
}
